KIPs/kip-113.md

```diff
@@ -0,0 +1,252 @@
+
+##### Usability from ServiceChain
+
+| Methods                      | [1] HeaderGov  | [2] ContractGov           | [3] SpareContract              | [4] ChainConfig | [5] Precompiled-contract          | [6] Bytecode-injection |
+| ---------------------------- | -------------- | ------------------------- | ------------------------------ | --------------- | --------------------------------- | ---------------------- |
+| Binary modification required | O              | O                         | X<br>Patch AddressBook address | O               | X<br> Patch the hardcoded address | O                      |
+| Contract deployment required | X<br> Registry | X <br> Registry, GovParam | X <br> Registry, AddressBook   | X<br> Registry  | X<br> Registry                    | X<br> AddressBook      |
+
+In terms of usability from ServiceChain, HeaderGov, ContractGov, ChainConfig, and Bytecode-injection are recommended.
+
+##### Others
+
+| Methods                                            | [1] HeaderGov                 | [2] ContractGov               | [3] SpareContract             | [4] ChainConfig      | [5] Precompiled-contract | [6] Bytecode-injection |
+| -------------------------------------------------- | ----------------------------- | ----------------------------- | ----------------------------- | -------------------- | ------------------------ | ---------------------- |
+| Immediate panic at hardfork from out-of-date nodes | △<br> At GovernanceData block | X<br> At the activation block | △<br> At SpareContract update | O                    | O                        | O                      |
+| Registry initial data setup before hardfork        | O                             | O                             | O                             | O                    | O                        | X                      |
+| Special address (like 0x0…400)                     | X<br> vanity address          | X<br> vanity address          | X<br> vanity address          | X<br> vanity address | O                        | O                      |
+
+In terms of various remaining aspects, ChainConfig, Precompiled-contract, and Bytecode-injection are recommended.
+
+Overall, we choose ChainConfig because
+
+- subscriber of Registry are mainly Klaytn nodes, not contracts so the awareness of subscriber contracts after address update is less important
+- ease of use in all of ServiceChain, Baobab (testnet), Cypress (mainnet)
+- initial data setup before hardfork is viable
+- special address format is less important
+
+## Backwards Compatibility
+
+This does not affect the backward compatibility as this is a newly deployed contract.
+
+## Implementation
+
+A reference implementation: [TODO: github link]
+
+### Cypress-specific implementation
+
+Cypress has the following access controls:
+
+- `registerPublicKey` reverts if `msg.sender` is not a `cnNodeId` from AddressBook
```

Comment:
```suggestion
- `registerPublicKey` reverts if `msg.sender` is not a valid `cnNodeId` registered in AddressBook.
```

There could be some dummy cnNodeId in the address book. Is this okay?

Reply:
Registry can't tell if it's a dummy or not, so it won't revert.